File: mac_robot_sim/Robot/robot.rb
# Copyright 2012, System Insights, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.

$: << '.'

require 'adapter'
require 'data_item'

include MTConnect

class Robot
  attr_reader :availability, :controller_mode, :execution, :material_load, :system
  attr_reader :material_unload, :open_chuck, :close_chuck
  attr_reader :open_door, :close_door, :adapter
  
  def initialize(port, cnc)
    @adapter = Adapter.new(port)
    @cnc = cnc

    @adapter.data_items << (@availability = DataItem.new('avail'))
    @adapter.data_items << (@system = SimpleCondition.new('system'))

    @adapter.data_items << (@controller_mode = DataItem.new('mode'))
    @adapter.data_items << (@execution = DataItem.new('exec'))


    @adapter.data_items << (@material_load = DataItem.new('material_load'))
    @adapter.data_items << (@material_unload = DataItem.new('material_unload'))

    @adapter.data_items << (@open_chuck = DataItem.new('open_chuck'))
    @adapter.data_items << (@close_chuck = DataItem.new('close_chuck'))

    @adapter.data_items << (@open_door = DataItem.new('open_door'))
    @adapter.data_items << (@close_door = DataItem.new('close_door'))
    
    @failed = false
  end

  def start
    @system.normal
    @adapter.start
    not_ready
  end
  
  def stop
    @adapter.stop
  end
  
  def gather(&block)
    @adapter.gather(&block)
  end
  
  def fail(value)
    @failed = value
  end
  
  def ready
    @adapter.gather do
      [@material_load, @material_unload, @open_chuck, @close_chuck,
      @open_door, @close_door].each { |di| di.value = "READY" }
    end
    @cnc.update_from_robot
  end
  
  def not_ready
    @adapter.gather do
      [@material_load, @material_unload, @open_chuck, @close_chuck,
      @open_door, @close_door].each { |di| di.value = "NOT_READY" }
    end
    @cnc.update_from_robot
  end
  
  def stop?
    @failed or !@system.normal?
  end
  
  def update_data_item(item, value)
    @adapter.gather do
      item.value = value
    end
    @cnc.update_from_robot    
  end
  
  def open_the_door
    update_data_item(@open_door, 'ACTIVE')
    sleep 1.5
    update_data_item(@open_door, 'READY')
  end
  
  def close_the_door
    update_data_item(@close_door, 'ACTIVE')
    sleep 1.5
    update_data_item(@close_door, 'READY')
  end
  
  def open_the_chuck
    update_data_item(@open_chuck, 'ACTIVE')
    sleep 1.5
    update_data_item(@open_chuck, 'READY')
  end
  
  def close_the_chuck
    update_data_item(@close_chuck, 'ACTIVE')
    sleep 1.5
    update_data_item(@close_chuck, 'READY')
  end
  
  def load_material
    if @cnc.cnc_material_load.stringValue != 'ACTIVE'
      update_data_item(@material_load, 'FAIL')
      sleep 2
      update_data_item(@material_load, 'READY')
      return
    end
    
    update_data_item(@material_load, 'ACTIVE')

    if @cnc.cnc_door_state.stringValue != 'OPEN'
      open_the_door
    end
    
    if @cnc.cnc_chuck_state.stringValue != 'OPEN'
      open_the_chuck
    end
    
    close_the_chuck
    
    sleep 1.5
    
    close_the_door
    
    sleep 2
    
    update_data_item(@material_load, 'COMPLETE')
    
    sleep 0.5
    
    update_data_item(@material_load, 'READY')
  end
  
  def unload_material
    if @cnc.cnc_material_unload.stringValue != 'ACTIVE' or
      @cnc.cnc_chuck_state.stringValue != 'CLOSED'
      update_data_item(@material_unload, 'FAIL')
      sleep 2
      update_data_item(@material_unload, 'READY')
      return
    end
    
    update_data_item(@material_unload, 'ACTIVE')
    
    if @cnc.cnc_door_state.stringValue != 'OPEN'
      open_the_door
    end
    
    sleep 1.5
    
    open_the_chuck
    
    sleep 2
    
    update_data_item(@material_unload, 'COMPLETE')
    
    sleep 0.5

    update_data_item(@material_unload, 'READY')
  end
end

File: mac_robot_sim/Robot/Robot.m
//
//  Robot.m
//  Robot
//
//  Created by William Sobel on 4/19/15.
//  Copyright (c) 2015 William Sobel. All rights reserved.
//

#import "Robot.h"
@import Foundation;

@implementation Robot

- initWithPort: (int) port andDelay: (int) delay {
  self = [super init];
  if (self != NULL) {
    _port = port;
    _failNext = false;
    _delay = delay;
  }
  return self;
}

- (void) notifyDelegate {
  if ([_delegate respondsToSelector: @selector(robotChanged:)])
    [_delegate robotChanged: self];
}

- (void) start {
  _adapter = [[MTConnectAdapter alloc] initWithPort: _port];
  
  [_adapter addDataItem: _avail = [[MTConnectDataItem alloc] initWithName: @"avail"]];
  [_adapter addDataItem: _mode = [[MTConnectDataItem alloc] initWithName: @"mode"]];
  [_adapter addDataItem: _exec = [[MTConnectDataItem alloc] initWithName: @"exec"]];
  [_adapter addDataItem: _system = [[MTConnectDataItem alloc] initWithName: @"system"]];
  [_adapter addDataItem: _material = [[MTConnectDataItem alloc] initWithName: @"material"]];
  [_adapter addDataItem: _materialLoad = [[MTConnectDataItem alloc] initWithName: @"material_load"]];
  [_adapter addDataItem: _materialUnload = [[MTConnectDataItem alloc] initWithName: @"material_unload"]];
  [_adapter addDataItem: _openChuck = [[MTConnectDataItem alloc] initWithName: @"open_chuck"]];
  [_adapter addDataItem: _closeChuck = [[MTConnectDataItem alloc] initWithName: @"close_chuck"]];
  [_adapter addDataItem: _openDoor = [[MTConnectDataItem alloc] initWithName: @"open_door"]];
  [_adapter addDataItem: _closeDoor = [[MTConnectDataItem alloc] initWithName: @"close_door"]];
  [_adapter addDataItem: _spindleInterlock = [[MTConnectDataItem alloc] initWithName: @"s_inter"]];
  [_adapter addDataItem: _chuckUnclamp = [[MTConnectDataItem alloc] initWithName: @"c_unclamp"]];
  
  _avail.value = @"AVAILABLE";
  _system.value = @"normal||||";
  _system.separateLine = YES;
  _mode.value = @"MANUAL";
  _exec.value = @"READY";
  _spindleInterlock.value = @"INACTIVE";
  _chuckUnclamp.value = @"INACTIVE";
  
  [self notReady];
  
  [_adapter start];
  [self notifyDelegate];
}

- (void) stop {
  [_adapter stop];
}

- (void) _update {
  [_adapter writeChangedDataItems];
  [self notifyDelegate];
}

- (void) update {
  if (_delay > 0.0) {
    NSTimer* _update = [NSTimer timerWithTimeInterval: _delay target: self
                                           selector: @selector(_update)
                                           userInfo: nil repeats: NO];
    [[NSRunLoop mainRunLoop] addTimer: _update forMode: NSDefaultRunLoopMode];
  } else {
    [self _update];
  }
}

- (void) receivedDataItem: (NSString*) name withValue: (NSString*) value {
  // Snake case the names and remove underscors _ OpenDoor NOT_READY -> openDoorNotReady
  NSString *prefix = [name stringByReplacingCharactersInRange: NSMakeRange(0,1)
                           withString: [[name substringToIndex: 1] lowercaseString]];
  NSString *suffix = [[value capitalizedString] stringByReplacingOccurrencesOfString: @"_" withString: @""];
  NSString *method = [prefix stringByAppendingString: suffix];
  
  // Get the selector and call method dynamically
  SEL selector = NSSelectorFromString(method);
  
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
  if ([self respondsToSelector: selector]) {
    if (_delay > 0.0) {
      NSTimer* disp = [NSTimer timerWithTimeInterval: _delay target: self
                                               selector: selector
                                               userInfo: nil repeats: NO];
      [[NSRunLoop mainRunLoop] addTimer: disp forMode: NSDefaultRunLoopMode];
    } else {
      [self performSelector: selector];
    }
  }
#pragma clang diagnostic pop
}

- (void) ready {
  _materialLoad.value = @"READY";
  _materialUnload.value = @"READY";
  _openChuck.value = @"READY";
  _closeChuck.value = @"READY";
  _openDoor.value = @"READY";
  _closeDoor.value = @"READY";
  [self update];
}

- (void) notReady {
  _materialLoad.value = @"NOT_READY";
  _materialUnload.value = @"NOT_READY";
  _openChuck.value = @"NOT_READY";
  _closeChuck.value = @"NOT_READY";
  _openDoor.value = @"NOT_READY";
  _closeDoor.value = @"NOT_READY";
  [self update];
}

- (void) fault {
  _system.value = @"fault|FAIL|||A fault occurred";
  [self update];
}

- (void) clearFault {
  _system.value = @"normal|||||";
  [self update];
}

- (void) loadTheMaterial {
  if (_failNext)
    _materialLoad.value = @"FAIL";
  else
    _materialLoad.value = @"ACTIVE";
  [self update];
  _failNext = false;

  [self openTheDoor];
}

- (void) materialLoadReady {
  if ([_materialLoad.value isEqualTo: @"COMPLETE"]) {
    _materialLoad.value = @"READY";
    [self update];
  }
}

- (void) materialLoadNotReady {
  [self materialLoadReady];
}

- (void) unloadTheMaterial {
  if (_failNext)
    _materialUnload.value = @"FAIL";
  else
    _materialUnload.value = @"ACTIVE";
  [self update];
  _failNext = false;
  
  [self openTheDoor];
}

- (void) materialUnloadReady {
  if ([_materialUnload.value isEqualTo: @"COMPLETE"])
  {
    _materialUnload.value = @"READY";
    [self update];
  }
}

- (void) materialUnloadNotReady {
  [self materialUnloadReady];
}


- (void) openTheDoor {
  _openDoor.value = @"ACTIVE";
  [self update];
}

- (void) openDoorComplete {
  _openDoor.value = @"READY";
  [self update];
  
  if ([_materialLoad.value isEqualTo: @"ACTIVE"] ||
      [_materialUnload.value isEqualTo: @"ACTIVE"]) {
    sleep(1);
    [self openTheChuck];
  }
}

- (void) closeTheDoor {
  _closeDoor.value = @"ACTIVE";
  [self update];
}

- (void) closeDoorComplete {
  _closeDoor.value = @"READY";
  [self update];
  
  if ([_materialLoad.value isEqualTo: @"ACTIVE"]) {
    _materialLoad.value = @"COMPLETE";
    [self update];
  }
}

- (void) openTheChuck {
  _openChuck.value = @"ACTIVE";
  [self update];
}

- (void) openChuckComplete {
  _openChuck.value = @"READY";
  [self update];
  
  if ([_materialLoad.value isEqualTo: @"ACTIVE"]) {
    sleep(1);
    [self closeTheChuck];
  } else if ([_materialUnload.value isEqualTo: @"ACTIVE"]) {
    _materialUnload.value = @"COMPLETE";
    [self update];
  }
}

- (void) closeTheChuck {
  _closeChuck.value = @"ACTIVE";
  [self update];
}

- (void) closeChuckComplete {
  _closeChuck.value = @"READY";
  [self update];
  
  if ([_materialLoad.value isEqualTo: @"ACTIVE"]) {
    sleep(1);
    [self closeTheDoor];
  }
}

#define MakeFail(interface) \
- (void) interface ## Fail { \
  if ([_ ## interface.value isEqualTo: @"ACTIVE"]) { \
    _ ## interface.value = @"FAIL"; \
    [self update]; \
  } \
} \
- (void) interface ## Ready { \
  if ([_ ## interface.value isEqualTo: @"FAIL"]) { \
    _ ## interface.value = @"READY"; \
    [self update]; \
  } \
} \
- (void) interface ## NotReady { [self interface ## Ready]; }

MakeFail(openDoor);
MakeFail(closeDoor);
MakeFail(closeChuck);
MakeFail(openChuck);

@end
